src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -2,59 +2,129 @@
-            [puppetlabs.puppetdb.query.environments :as environments]
             [puppetlabs.puppetdb.query.events :as events]
             [puppetlabs.puppetdb.query.event-counts :as event-counts]
             [puppetlabs.puppetdb.query.facts :as facts]
             [puppetlabs.puppetdb.query.factsets :as factsets]
             [puppetlabs.puppetdb.query.fact-contents :as fact-contents]
-            [puppetlabs.puppetdb.query.nodes :as nodes]
             [puppetlabs.puppetdb.query.reports :as reports]
             [puppetlabs.puppetdb.query.report-data :as report-data]
             [puppetlabs.puppetdb.query.resources :as resources]
             [puppetlabs.puppetdb.scf.storage-utils :as su]
             [puppetlabs.puppetdb.schema :as pls]
+            [puppetlabs.puppetdb.query-eng.engine :as eng]
             [schema.core :as s]))
 
-(defn entity->sql-fns
+(def entity-fn-idx
+  (atom
+    {:aggregate-event-counts {:munge aggregate-event-counts/munge-result-rows
+                              :rec nil}
+     :event-counts {:munge event-counts/munge-result-rows
+                    :rec nil}
+     :facts {:munge facts/munge-result-rows
+             :rec eng/facts-query}
+     :fact-contents {:munge fact-contents/munge-result-rows
+                     :rec eng/fact-contents-query}
+     :fact-paths {:munge facts/munge-path-result-rows
+                  :rec eng/fact-paths-query}
+     :fact-names {:munge facts/munge-name-result-rows
+                  :rec eng/fact-names-query}
+     :factsets {:munge factsets/munge-result-rows
+                :rec eng/factsets-query}
+     :catalogs {:munge catalogs/munge-result-rows
+                :rec eng/catalog-query}
+     :nodes {:munge (constantly identity)
+             :rec eng/nodes-query}
+     :environments {:munge (constantly identity)
+                    :rec eng/environments-query}
+     :events {:munge events/munge-result-rows
+              :rec eng/report-events-query}
+     :edges {:munge edges/munge-result-rows
+             :rec eng/edges-query}
+     :reports {:munge reports/munge-result-rows
+               :rec eng/reports-query}
+     :report-metrics {:munge (report-data/munge-result-rows :metrics)
+                      :rec eng/report-metrics-query}
+     :report-logs {:munge (report-data/munge-result-rows :logs)
+                   :rec eng/report-logs-query}
+     :resources {:munge resources/munge-result-rows
+                 :rec eng/resources-query}}))
+
+(defn assoc-in-idx!
+  "assoc into the entity index as if query recs were maps"
+  [[entity & ks] v]
+  (if (and (= :rec (first ks)) (next ks))
+    ;; if it's a query rec and we're updating a nested component, call the
+    ;; existing rec and wrap val in a function
+    (let [rec' (-> ((get-in @entity-fn-idx [entity :rec]))
+                   (assoc-in (next ks) v))]
+      (swap! entity-fn-idx #(assoc-in % [entity :rec] (fn [] rec'))))
+    (swap! entity-fn-idx #(assoc-in % (cons entity ks) v))))
```

Comment:
I think the `#(` is redundant you can just use `(swap! entity-fn-idx assoc-in (cons entity ks) v)))`